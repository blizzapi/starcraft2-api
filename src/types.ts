import {
  BlizzAPIOptions,
  RegionIdAsNumberOrString,
} from 'blizzapi';

export interface StarCraft2APIOptions extends BlizzAPIOptions {}

export interface League {
  seasonId: number | string;
  queueId: number | string;
  teamType: number | string;
  leagueId: number | string;
}

export interface PlayerObject {
  regionId: RegionIdAsNumberOrString;
  realmId: number | string;
  profileId: number | string;
  locale?: string;
}
