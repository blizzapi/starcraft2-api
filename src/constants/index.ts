export { profileUrlLocales as PROFILE_URL_LOCALES } from "./profileUrlLocales";
