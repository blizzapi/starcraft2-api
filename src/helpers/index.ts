export { checkIfProfileIdLooksValid } from "./profileIds";
export {
  checkIfProfileUrlLocaleLooksValid,
  getAllProfileUrlLocales,
  profileUrlLocaleRegex,
  validateProfileUrlLocale,
} from "./profileUrlLocales";
export {
  checkIfProfileUrlLooksValid,
  constructProfileUrl,
  profileUrlRegex,
  unpackProfileUrl,
  validateProfileUrl,
} from "./profileUrls";
