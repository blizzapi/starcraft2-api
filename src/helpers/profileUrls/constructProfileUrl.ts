import BlizzAPI from 'blizzapi';
import checkIfProfileIdLooksValid from '../profileIds/checkIfProfileIdLooksValid';
import validateProfileUrlLocale from '../profileUrlLocales/validateProfileUrlLocale';
import {
  PlayerObject,
} from '../../types';

export default (
  {
    regionId,
    realmId,
    profileId,
  }: PlayerObject,
  localeName?: string,
) => {
  try {
    const validRegionId = BlizzAPI.validateRegionId(regionId);
    const validRealmId = BlizzAPI.checkIfSc2RealmLooksValid(realmId);
    const validProfileId = checkIfProfileIdLooksValid(profileId);
    const locale = localeName && validateProfileUrlLocale(localeName)
      ? localeName
      : 'en-us';

    if (validRegionId && validRealmId && validProfileId) {
      return `https://starcraft2.com/${locale}/profile/${regionId}/${realmId}/${profileId}`;
    }
    return '';
  } catch (error) {
    return '';
  }
};
