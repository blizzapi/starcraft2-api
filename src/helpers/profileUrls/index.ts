export { checkIfProfileUrlLooksValid } from "./checkIfProfileUrlLooksValid";
export { constructProfileUrl } from "./constructProfileUrl";
export { profileUrlRegex } from "./profileUrlRegex";
export { unpackProfileUrl } from "./unpackProfileUrl";
export { validateProfileUrl } from "./validateProfileUrl";
