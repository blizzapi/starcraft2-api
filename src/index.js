const Profile = require('./api/profile');
const Ladder = require('./api/ladder');
const Account = require('./api/account');
const Legacy = require('./api/legacy');

class StarCraft2API {
    constructor(clientID, clientSecret) {
        this.clientID = clientID;
        this.clientSecret = clientSecret;
    }
    
    Profile() {
        return new Profile(this.clientID, this.clientSecret);
    }

    Ladder() {
        return new Ladder(this.clientID, this.clientSecret);
    }

    Account() {
        return new Account(this.clientID, this.clientSecret);
    }

    Legacy() {
        return new Legacy(this.clientID, this.clientSecret);
    }
}

module.exports = StarCraft2API;
