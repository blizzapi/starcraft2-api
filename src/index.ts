import { BlizzAPI, BlizzUtils } from 'blizzapi';
import { QueryOptions } from 'blizzapi/dist/lib/interfaces';
import {
  ClientSecret,
  AccessToken,
  ClientId,
  RegionIdOrName,
  RegionIdAsNumberOrString,
  Locale,
  Sc2Realm,
} from 'blizzapi/dist/lib/types';

// tslint:disable no-class no-expression-statement no-this
export default class StarCraft2API extends BlizzAPI {
  constructor(
    region: RegionIdOrName,
    clientId: ClientId,
    clientSecret: ClientSecret,
    accessToken?: AccessToken,
    options?: QueryOptions,
  ) {
    super(region, clientId, clientSecret, accessToken, options);
  }

  queryStaticProfileData(regionId: RegionIdAsNumberOrString, locale?: Locale): Promise<object> {
    const queryLocale = locale || BlizzUtils.getDefaultLocaleNameForRegionId(regionId);
    return this.query(`/sc2/static/profile/${regionId}?locale=${queryLocale}`);
  }

  queryProfileMetadata(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
    locale?: Locale,
  ): Promise<object> {
    const queryLocale = locale || BlizzUtils.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/metadata/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`,
    );
  }

  queryProfile(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
    locale?: Locale,
  ): Promise<object> {
    const queryLocale = locale || BlizzUtils.getDefaultLocaleNameForRegionId(regionId);
    return this.query(`/sc2/profile/${regionId}/${realmId}/${profileId}?locale=${queryLocale}`);
  }

  queryLadderSummary(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
    locale?: Locale,
  ): Promise<object> {
    const queryLocale = locale || BlizzUtils.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/summary?locale=${queryLocale}`,
    );
  }

  queryPlayerLadder(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
    ladderId: number | string,
    locale?: Locale,
  ): Promise<object> {
    const queryLocale = locale || BlizzUtils.getDefaultLocaleNameForRegionId(regionId);
    return this.query(
      `/sc2/profile/${regionId}/${realmId}/${profileId}/ladder/${ladderId}?locale=${queryLocale}`,
    );
  }

  queryGrandmasterLeaderboard(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/ladder/grandmaster/${regionId}`);
  }

  querySeason(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/ladder/season/${regionId}`);
  }

  queryPlayerAccount(accountId: number | string): Promise<object> {
    return this.query(`/s2/player/${accountId}`);
  }

  queryLegacyProfile(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
  ): Promise<object> {
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}`);
  }

  queryLegacyLadders(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
  ): Promise<object> {
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/ladders`);
  }

  queryLegacyMatchHistory(
    regionId: RegionIdAsNumberOrString,
    realmId: Sc2Realm,
    profileId: number | string,
  ): Promise<object> {
    return this.query(`/sc2/legacy/profile/${regionId}/${realmId}/${profileId}/matches`);
  }

  queryLegacyLadder(
    regionId: RegionIdAsNumberOrString,
    ladderId: number | string,
  ): Promise<object> {
    return this.query(`/sc2/legacy/ladder/${regionId}/${ladderId}`);
  }

  queryLegacyAchievements(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/legacy/data/achievements/${regionId}`);
  }

  queryLegacyRewards(regionId: RegionIdAsNumberOrString): Promise<object> {
    return this.query(`/sc2/legacy/data/rewards/${regionId}`);
  }
}
// tslint:enable no-class no-expression-statement
