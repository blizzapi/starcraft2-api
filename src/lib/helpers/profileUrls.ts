import BlizzAPI, {
  RegionIdAsNumberOrString,
  Sc2RealmAsNumberOrString,
} from 'blizzapi';

export interface PlayerProfile {
  regionId: RegionIdAsNumberOrString,
  realmId: Sc2RealmAsNumberOrString,
  profileId: string | number,
}

const profileUrlRegex: RegExp =
  /starcraft2\.com\/([A-z]{2}-[A-z]{2})\/profile\/(1|2|3|5{1})\/([0-9]{1})\/([0-9]*)/gi;

export const validateProfileUrl = (url: string) =>
  profileUrlRegex.test(url);

export const validateProfileId = (profileId: number | string) => 
  profileId.toString() === parseInt(profileId as string, 10).toString();

export const unpackProfileUrl = (url: string, includeLocale?: boolean) => {
  const urlIsValid = validateProfileUrl(url);

  if (!urlIsValid) return {};

  const profileUrl = url.match(profileUrlRegex)![0];
  const profileDataArray = profileUrlRegex.exec(profileUrl)!;

  // tslint:disable-next-line: no-object-mutation
  profileUrlRegex.lastIndex = 0;

  const playerObject = {
    regionId: profileDataArray[2],
    realmId: profileDataArray[3],
    profileId: profileDataArray[4],
  };

  return includeLocale
    ? {
      locale: profileDataArray[1],
      ...playerObject,
    }
    : playerObject;
}

export const constructProfileUrl = ({ regionId, realmId, profileId }: PlayerProfile, localeName?: string) => {
  try {
    const validRegionId = BlizzAPI.validateRegionId(regionId);
    const validProfileId = validateProfileId(profileId);
    const validRealmId = BlizzAPI.checkIfSc2RealmLooksValid(realmId);
  
    if (validRegionId && validRealmId && validProfileId) {
      const locale = localeName || 'en-us';
      return `https://starcraft2.com/${locale}/profile/${regionId}/${realmId}/${profileId}`;
    }
    return '';
  } catch (error) {
    return '';
  }
}