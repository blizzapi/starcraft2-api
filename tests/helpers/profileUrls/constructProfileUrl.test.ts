import { constructProfileUrl } from "../../../src/helpers/profileUrls/constructProfileUrl";
import playerObjects from "../../__testData__/playerObjects.json";
import invalidPlayerObjects from "../../__testData__/invalidPlayerObjects.json";
import { PlayerObject } from "../../../src/types";

describe("constructProfileUrl", () => {
  (playerObjects as PlayerObject[]).forEach((playerObject) => {
    it("should match snapshot for valid player object", () => {
      expect(constructProfileUrl(playerObject)).toMatchSnapshot();
    });
  });

  invalidPlayerObjects.forEach((invalidPlayerObject) => {
    it("should match snapshot for invalid player object", () => {
      expect(
        constructProfileUrl(invalidPlayerObject as PlayerObject)
      ).toMatchSnapshot();
    });
  });

  (playerObjects as PlayerObject[]).forEach((playerObject) => {
    it("should match snapshot for valid player object and valid locale", () => {
      expect(constructProfileUrl(playerObject, "en-us")).toMatchSnapshot();
    });
  });

  (playerObjects as PlayerObject[]).forEach((playerObject) => {
    it("should match snapshot for valid player object and invalid locale", () => {
      expect(
        constructProfileUrl(playerObject, "invalidLocale")
      ).toMatchSnapshot();
    });
  });
});
