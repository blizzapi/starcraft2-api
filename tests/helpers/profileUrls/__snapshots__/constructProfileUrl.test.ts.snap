// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructProfileUrl should match snapshot for invalid player object 1`] = `""`;

exports[`constructProfileUrl should match snapshot for invalid player object 2`] = `""`;

exports[`constructProfileUrl should match snapshot for invalid player object 3`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object 1`] = `"https://starcraft2.com/en-us/profile/1/1/12453456"`;

exports[`constructProfileUrl should match snapshot for valid player object 2`] = `"https://starcraft2.com/en-us/profile/2/2/34987654"`;

exports[`constructProfileUrl should match snapshot for valid player object 3`] = `"https://starcraft2.com/en-us/profile/3/2/12987654"`;

exports[`constructProfileUrl should match snapshot for valid player object 4`] = `"https://starcraft2.com/en-us/profile/1/1/12453456"`;

exports[`constructProfileUrl should match snapshot for valid player object 5`] = `"https://starcraft2.com/en-us/profile/2/2/34987654"`;

exports[`constructProfileUrl should match snapshot for valid player object 6`] = `"https://starcraft2.com/en-us/profile/3/2/12987654"`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 1`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 2`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 3`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 4`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 5`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and invalid locale 6`] = `""`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 1`] = `"https://starcraft2.com/en-us/profile/1/1/12453456"`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 2`] = `"https://starcraft2.com/en-us/profile/2/2/34987654"`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 3`] = `"https://starcraft2.com/en-us/profile/3/2/12987654"`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 4`] = `"https://starcraft2.com/en-us/profile/1/1/12453456"`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 5`] = `"https://starcraft2.com/en-us/profile/2/2/34987654"`;

exports[`constructProfileUrl should match snapshot for valid player object and valid locale 6`] = `"https://starcraft2.com/en-us/profile/3/2/12987654"`;
